### ISSUE TITLE
How does ClearML compare to MLRun?

### ISSUE THREAD

**Issue by Make42** [allegroai/clearml#718]
We are currently product-hunting for our MLOps infrastructure and ClearML, Kedro, MLRun are on our short list. **How does ClearML compare to MLRun?** One big difference seems to be that MLRun has a feature store integrated. What are advantages/disadvantages of each, compared to each other?

**Comment by Make42** [comment]
Hi @erezalg, could you add links as evidence for your respective claims into your post?
- E.g., where did you learn that the "serverless functions" are proprietary? I had the impression that they are part of FOSS MLRun. The online serving with Nuclio also seems to be open source to me.
- Also, where did you get from that MLRun only support tabular data?

According to the MLRun proponents, MLRun is strong on execution, including scaling and online serving. But basically they also claim to be end-to-end: https://github.com/mlrun/mlrun/issues/2120

Interesting, both MLRun and ClearML proponents argue similarly: ClearML's said that it has better community in terms of responsiveness at Slack etc. and MLRun's said that they have more github commits.

-----------------------

ClearML does not seem to have data monitoring (meaning things like drift detection in data) - after all it only manages data file in a black-box manner - which comes with the data agnosticism. This is partly because of the lack of a feature store, I guess.

I am not sure whether ClearML has hooks into Git repositories. I would like to have something that checks out my commit and runs a specific script, once I commit to a certain branch.

**Comment by erezalg** [comment]
Hi @Make42,

First, I'll state the obvious that everything I get here is only from reading their docs. I'm by no means expert on mlrun and I might not get it all correctly :)

 - About the serverless, as I understand, they rely on nuclio (which is written by the same company). Maybe proprietary was not the right word :) I meant that they use Nuclio for executing code which is their own serverless technology
 -  Looking at their examples, they only have XGBoost and Sklearn one. I now saw that on their website they have integration to TF and pytorch, but saw only using model files of those and not tracking experiments. But again, I might just have missed that :) 
 
About both being end-to-end yeah that makes sense, the two platforms might indeed be end-to-end with different emphasis.

ClearML doesn't implement drift detection algorithms itself, but we use Nvidia triton for serving and I'm not 100% sure what it supports. You can check [here](https://github.com/thepycoder/asteroid_example) a repository that implemented drift detection with grafana visualizations for drift detection with ClearML serving. We indeed don't, at the moment, have drift detection at the data management point. 

As for community, I mean I guess we're both right? It depends on what you're focusing on and what's important to you.

As for Git integration, ClearML captures that git information when integrated (repo \ branch \ commit \ tag) and can rerun the commit when re-running with ClearML Agent.

**Comment by gilad-shaham** [comment]
@Make42 to comment on the mlrun side, besides the answer at https://github.com/mlrun/mlrun/issues/2120#issuecomment-1183640109

The serverless portion is not just for serving, serverless essentially mean you don't have to worry about resource provisioning and scale. MLRun provides the user (data scientist in most cases) this ability for anything it executes, including jobs. For example, one can run Dask or Horovod (MPI) for distributed training, all the user has to do is define the number of CPUs/GPUs and MLRun will execute that (immediately or scheduled). Indeed for serving MLRun leverages Nuclio which is an open source (Apache 2.0 license) with 4.5K stars on GitHub and it is a well established cross-platform serverless function library. The strengths of Nuclio allows us to run at scale and realtime, it just made very little sense to re-implement this when Nuclio solves the issue. However, Nuclio is just the underlying technology, MLRun adds model serving classes and realtime data pipeline execution which goes way beyond the Nuclio serverless function capabilities.

As for PyTorch, TensorFlow and supporting unstructured data, these are supported and we have examples that show images classification and NLP. You can find those in the [demos repo](https://github.com/mlrun/demos). MLRun can execute on GPUs for both training and serving. With regards to experiment tracking, we preferred to use TensorBoard rather than reimplement it. See also [a recent blog post](https://www.iguazio.com/blog/machine-learning-experiment-tracking-from-zero-to-hero-in-2-lines-of-code/) regarding auto logging with these libraries.

As for feature store, indeed MLRun is one of the few open source libraries to have a feature store. Note that it is not mandatory to use, it's yet another capability that you can choose to leverage, depending on your use case. The advantage of having it available is that you don't have to handle all the integration of another library in order to start using the feature store.

Hope it helps

**Comment by erezalg** [comment]
Hi Everyone! I think this thread has veered a bit of-topic for a github issue about ClearML :smile: 
We're removing this thread and you can take it into DM's or mlrun's github repo.

Cheers!
